refactor(MembershipsSdkAdapter): sort the members in a room

CernaianuMihai opened the pull request:
https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-223396

cipak commented on src/MembershipsSDKAdapter.js:
```diff
@@ -117,11 +136,17 @@ export default class MembershipsSDKAdapter extends MembershipsAdapter {
-      map((page) => page.items.map(membershipToMember)),
+    }));
+
+    const orderedMembers$ = currentUser$.pipe(
```
try using [combineLatest](https://rxjs-dev.firebaseapp.com/api/index/function/combineLatest)

CernaianuMihai replied:
Apparently combineLatest is deprecated and it recommends me to use pipe to map instead
![deprecated_combineLatest](https://user-images.githubusercontent.com/56340398/114694021-d6047f80-9d22-11eb-961a-825007c86c0e.png)

cipak replied:
I think the `resultSelector` parameter is deprecated, not the operator itself. Maybe you called the operator like this `combineLatest($a, $b)` instead of `combineLatest([$a, $b])`?

CernaianuMihai replied:
Yes, i just tried calling the operator like this combineLatest([$a, $b]), but it's still deprecated
![deprecated_combine](https://user-images.githubusercontent.com/56340398/114834223-0575c380-9dd9-11eb-8ddc-51056f682c08.png)

cipak commented on src/MembershipsSDKAdapter.js:
```diff
@@ -117,11 +136,17 @@ export default class MembershipsSDKAdapter extends MembershipsAdapter {
       inMeeting: null,
     });
 
+    const currentUser$ = new PeopleSDKAdapter(this.datasource).getMe();
```
I don't like the fact that a new instance of people's adapter is created. We should be able to use the same one. However, it's a "sibling" adapter, I'm not sure if there's currently any way to access it. @lalli-flores

cipak replied:
I'm thinking of adding a "link" to the `WebexSDKAdapter` instance from each specific adapter instance. This could be done by sending an extra parameter to each adapter constructor:
```js
this.peopleAdapter = new PeopleSDKAdapter(sdk, this);
```
https://github.com/webex/sdk-component-adapter/blob/master/src/WebexSDKAdapter.js#L16

Having this, we could easily access sibling adapters:
```js
const me$ = this.parent.peopleAdapter.getMe(); // or this.main, this.adapter, this.adapters, this.siblingAdapters, this.siblings, ...
```

cipak commented on src/MembershipsSDKAdapter.js:
```diff
@@ -119,11 +146,17 @@ export default class MembershipsSDKAdapter extends MembershipsAdapter {
+      map((page) => page.items),
+    );
+
+    const orderedMembers$ = combineLatest([me$, memberships$]).pipe(
```
you can now rename `orderedMembers$` to `members$` , to revert some of the changes below.

cipak commented on src/MembershipsSDKAdapter.js:
```diff
@@ -119,11 +146,17 @@ export default class MembershipsSDKAdapter extends MembershipsAdapter {
     });
 
-    const members$ = defer(() => this.datasource.memberships.list({
+    const me$ = from(this.datasource.people.get('me'));
```
`from()` should be better than `defer()` in this case, because we don't want the `get(me)` call to be performed on each membership event.

lalli-flores reviewed (commented):
The code looks good, but tests need to be updated properly.
Update test description to `emits a sorted member list on subscription`.
You're also missing tests for the `sortRoomMembers` function

lalli-flores commented:
> You're also missing tests for the sortRoomMembers function

CernaianuMihai commented:
I think the test for `getMembersFromDestionation` is enough because `sortRoomMembers` is called inside of that test and it's an internal function.

lalli-flores commented:
I disagree. I understand that by virtue of using `getMembersFromDestionation`, `sortRoomMembers` is also called but that doesn't make it "tested".

Adding tests for `sortRoomMembers` allows us to document code by explicitly stating what scenarios we expect `sortRoomMembers` to handle. This is useful for new developers that are trying to get familiar with the code base, and for contributors that haven't looked at the code in a while.
This is also helpful for debugging, where output from `getMembersFromDestionation` might not be showing up as expected, and by having separate tests for `sortRoomMembers` we can quickly assess that is not because an issue there.

In terms of testing strategy, helpers like `sortRoomMembers` should be tested with mock data (which doesn't have to be associated with the overall SDK mock), while for public methods we move towards integration-type tests. I would like to start aligning towards this goal.

cipak commented:
While I can see the benefit of new developers looking at unit tests to get a better understanding of the existing code base, I feel that the disadvantages of unit testing non-public methods outweigh that benefit:
- more difficult "internal guts" refactoring (unit tests need to be updated as well)
- after such an "internal guts" refactoring, the unit tests no longer tell us we did a good job (they fail anyway)
- test duplication, the same or similar scenarios need to be applied at public level but also private level functions
- we are testing the internal implementation instead of just the public contract (api) of a module
- devs will be tempted to avoid creating a new function. In our case, we can easily inline the sorting in the `getRoomMembers` function to avoid the extra tests, which leads to less clear code. (Btw, `getRoomMembers` does not have unit tests, and neither do `get(Active)Members` and all the other non-public functions in this module)

CernaianuMihai commented:
We will also have to export the `sortRoomMembers` function only for testing it.

CernaianuMihai commented:
@lalli-flores Anything else that need to be done in this PR or can it be merged?

lalli-flores commented:
If you don't think it's a good idea to test `sortRoomMembers`, that's ok, but you at least have to make sure that the scenarios are covered by the public API. I don't see a test for showing when there is the "current" user or when there persons with no display name. There should be 3 scenarios tested total.

CernaianuMihai commented:
@lalli-flores I added a person with empty name and a person that matches with the current user in mockData and I updated the 'emits a sorted member list on subscription' test to test those two as well.

lalli-flores reviewed (approved):
Thanks for updating the tests @CernaianuMihai

webex-components-publisher commented:
:tada: This PR is included in version 1.33.2 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/@webex/sdk-component-adapter/v/1.33.2)
- [GitHub release](https://github.com/webex/sdk-component-adapter/releases/tag/v1.33.2)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
